File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int N,M;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        input();

        int answer = binary_search();

        System.out.println(answer);
    }

    public static int binary_search(){
        int start = 0 ;
        int end = Arrays.stream(arr).max().getAsInt();

        while(start + 1 < end){
            int mid = (start + end)/2;
            if(check(mid)) start = mid;
            else end = mid;
        }

        return start;
    }

    public static boolean check(int mid){
        long sum =0;
        for(int i=0; i<N; i++){
            if(arr[i]>mid) sum+= arr[i]-mid;
        }

        return sum>=M;
    }

    public static void input() throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for(int n=0; n<N; n++){
            arr[n] = Integer.parseInt(st.nextToken());
        }
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
def check(mid):
    tree_sum = 0
    for h in lst:
        if h>mid:
            tree_sum += h-mid
    if tree_sum >= m:
        return True
    return False
n, m = map(int, input().split())
lst = list(map(int, input().split()))

st = 0
ed = max(lst)
answer = 0
while st<=ed:
    mid = (st+ed)//2
    if check(mid):
        answer = mid
        st = mid+1
    else:
        ed = mid-1
print(answer)